### ISSUE TITLE
Inline initialization and merging of dicts to strict

### ISSUE THREAD

**Issue by Kaniee** [microsoft/pyright#5570]
**Describe the bug**
When I create two dicts and merge them in the same statement pyright complains about a invariant dict parameter.
If I create the two dicts separately and merge them afterwards no errors are thrown.

**Code or Screenshots**
```python
result: dict[str, float] = {"a": 2} | {"b": 2} # Error

a: dict[str, float] = {"a": 2}
b: dict[str, float] = {"b": 2}
result: dict[str, float] = a | b # No error

result: dict[str, float] = {**{"a": 2}, **{"b": 2}} # No error
```

**Output**
```
test_file.py1:28 - error: Expression of type "dict[str, int]" cannot be assigned to declared type "dict[str, float]"
     "dict[str, int]" is incompatible with "dict[str, float]"
           Type parameter "_VT@dict" is invariant, but "int" is not the same as "float" (reportGeneralTypeIssues)
```

**

**VS Code extension or command-line**
`pyright 1.1.318`
`Python 3.10.6`

**Comment by erictraut** [comment]
This behavior is as expected. Binary operators like `|` are complex in Python. They are not a simple application of a single magic method. These operations involve pairs of magic methods that are used at runtime. In the case of `|`, this includes `__or__` and `__ror__`. If the operand types are unions, this gets even more complicated because some combinations of subtypes might use `__or__` while others use `__ror__`. To evaluate the effects of such an operator, a type checker needs to first determine both the left and right side (both operands) and then determine whether to use `__or__` or `__ror__`. That means bidirectional type inference cannot be used in these cases. Normal type inference rules therefore apply when evaluating expressions like `{"a": 2}` and `{"b": 2}` in the first line of your example. Both of these expressions are inferred as `dict[str, int]`. Applying the operator `|` produces `dict[str, int]`, which is incompatible with `dict[str, float]`. 

In your "no error" example, bidirectional type inference is applied, and this overrides the normal type inference rules. Therefore, `{"a": 2}` is inferred as `dict[str, float]` because of the "expected type" supplied by assignment target.

I understand why you consider this an undesirable behavior, but it is expected given how type evaluation and type inference work in pyright. I don't consider this a bug.
